### PR TITLE
Refactor the logic of Reorder and Project

### PR DESCRIPTION
**Pull request by HangyuanLiu**
1. Eliminate projections in the logical phase, use LogicalProjectOperator
2. Refactor JoinAssociativityRule
3. Refactor JoinOrder. Before entering the Reorder algorithm, treat Project+Join as a group, which is to ensure that they are reordered as a whole

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java [hunk 1]
```diff
@@ -206,10 +206,7 @@ private void mergeGroup(Group srcGroup, Group dstGroup) {
         // When some rule merge two groups to one group, or
         // the GroupExpressions of one group are all removed.
         // The group is empty, We should remove it.
-        List<Group> groups = getAllEmptyGroups();
-        for (Group group : groups) {
-            removeOneGroup(group);
-        }
+        removeAllEmptyGroup();
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:209]: Why change this?

**Comment by kangkaisen** [comment]
Need comment why refactor and how refactor

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java (hunk 1 above)
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:209]: In the original way, a new empty group may appear, resulting in an error in the subsequent binder

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentTest.java [hunk 2]
```diff
@@ -404,6 +404,7 @@ public void testLeftOuterJoinWithPredicate() throws Exception {
     public void testCrossJoinToInnerJoin() throws Exception {
         String sql = "SELECT t0.v1 from t0, test_all_type where t0.v1 = test_all_type.t1d";
         String planFragment = getFragmentPlan(sql);
+        System.out.println(planFragment);
```
Comment (Seaven) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/PlanFragmentTest.java:407]: remove, and other too

**Comment by HangyuanLiu** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java [hunk 3]
```diff
@@ -387,11 +387,7 @@ public LogicalPlan visitTable(TableRelation node, ExpressionMapping context) {
 
         OptExprBuilder scanBuilder = new OptExprBuilder(scanOperator, Collections.emptyList(),
                 new ExpressionMapping(node.getScope(), outputVariables));
-        LogicalProjectOperator projectOperator =
-                new LogicalProjectOperator(outputVariables.stream().distinct()
-                        .collect(Collectors.toMap(Function.identity(), Function.identity())));
-
-        return new LogicalPlan(scanBuilder.withNewRoot(projectOperator), outputVariables, null);
+        return new LogicalPlan(scanBuilder, outputVariables, null);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/RelationTransformer.java:390]: Why remove this project

**Pull request closed without merging**